/**:
  ros__parameters:
    ############ system configuration ############
    # Data logging
    log_to_file: true
    log_debug: false
    log_enabled: [
        ## ADD LOGGER IDS HERE
        # tactic
        "tactic",
        "tactic.module",
        # "tactic.module.live_mem_manager",
        # pose graph
        # "pose_graph",
        # pipeline specific
        # "lidar.pipeline",
        # "lidar.velodyne_converter",
        # "lidar.honeycomb_converter",
        # "lidar.preprocessing",
        # "lidar.odometry_map_recall",
        # "lidar.odometry_icp",
        # "lidar.odometry_map_merging",
        # "lidar.keyframe_test",
        "lidar.localization_map_recall",
        "lidar.localization_icp",
        "lidar.intra_exp_merging",
        "lidar.dynamic_detection",
        "lidar.inter_exp_merging",
      ]

    ############ tactic configuration ############
    tactic:
      task_queue_num_threads: 2
      task_queue_size: -1

      extrapolate_odometry: true # path tracker
      localization_only_keyframe: false
      localization_skippable: false
      default_loc_cov: [1.0, 0.5, 0.5, 0.25, 0.25, 0.5]
      merge_threshold: [5.0, 3.0, 0.2]

      visualize: true
      vis_loc_path_offset: [0., 0., -10.]

      chain:
        min_cusp_distance: 1.5
        angle_weight: 7.0
        search_depth: 5
        search_back_depth: 10
        distance_warning: 5.0

    ############ pipeline configuration ############
    pipeline:
      type: lidar

      preprocessing: [
          # preprocessing thread modules
          "conversion",
          "filtering",
        ]

      odometry: [
          # odometry and mapping thread modules
          "recall",
          "icp",
          "mapping",
          "keyframe_test",
          # "intra_exp_merging",
          # "dynamic_detection",
          # "inter_exp_merging",
          "memory",
        ]

      localization: [
          # localization thread modules
          "recall",
          "icp",
          "memory",
        ]

    preprocessing:
      conversion:
        type: lidar.honeycomb_converter_v2

        visualize: true

      filtering:
        type: lidar.preprocessing_v2

        num_threads: 4
        frame_voxel_size: 0.1 # grid subsampling voxel size

        vertical_angle_res: 0.0132645 # vertical angle resolution in radius, equal to 0.76 degree documented in the manual
        polar_r_scale: 2.0 # polar_r_scale x vertical_angle_res = nearest neighbor search radius for normal computation
        r_scale: 4.0 # scale down point range by this value after taking log, whatever works
        h_scale: 1.54 # scale down yaw(phi) by this value so that vertical resolution ~= horizontal resolution, horizontal resolution when 5Hz spin frequence is ~1.17 degree, so 1.17 / 0.76 = 1.54

        num_sample1: 10000 # max number of sample after filtering based on planarity
        min_norm_score1: 0.95 # min planarity score

        num_sample2: 10000 # max number of sample after filtering based on planarity
        min_norm_score2: 0.2 # 0.2 is when the incident angle 5/12 * pi
        min_normal_estimate_dist: 2.0 # minimum distance to estimate normal in meters
        max_normal_estimate_angle: 0.42 # must <1/2, this value will be timed by M_PI

        cluster_num_sample: 10000 # maxnumber of sample after removing isolated points

        visualize: true

    odometry:
      recall:
        type: lidar.odometry_map_recall

        visualize: true

      icp:
        type: lidar.odometry_icp_v2

        # Prior terms
        trajectory_smoothing: true
        use_constant_acc: false
        lin_acc_std_dev_x: 1.0
        lin_acc_std_dev_y: 0.1
        lin_acc_std_dev_z: 0.1
        ang_acc_std_dev_x: 0.1
        ang_acc_std_dev_y: 0.1
        ang_acc_std_dev_z: 1.0

        # ICP parameters
        num_threads: 4
        first_num_steps: 3
        initial_max_iter: 6
        initial_num_samples: 12000
        initial_max_pairing_dist: 1.5
        initial_max_planar_dist: 1.0
        refined_max_iter: 50
        refined_num_samples: 12000
        refined_max_pairing_dist: 1.0
        refined_max_planar_dist: 0.3

        min_matched_ratio: 0.5

        visualize: true

      mapping:
        type: lidar.odometry_map_merging

        map_voxel_size: 0.3

        visualize: true

      keyframe_test:
        type: lidar.keyframe_test

        min_translation: 1.0
        min_rotation: 15.0
        max_translation: 3.0
        max_rotation: 30.0

        min_matched_points_ratio: 0.6

        max_num_points: 100000

      intra_exp_merging:
        type: "lidar.intra_exp_merging"
        depth: 2

        visualize: true

      dynamic_detection:
        type: "lidar.dynamic_detection"
        depth: 2

        horizontal_resolution: 0.02042
        vertical_resolution: 0.01326
        max_num_observations: 2000
        min_num_observations: 4
        dynamic_threshold: 0.3

        visualize: true

      inter_exp_merging:
        type: "lidar.inter_exp_merging"
        depth: 2

        horizontal_resolution: 0.02042
        vertical_resolution: 0.01326
        max_num_observations: 2000
        min_num_observations: 4
        dynamic_threshold: 0.3

        max_num_experiences: 128

        visualize: true

      memory:
        type: "live_mem_manager"
        window_size: 5

    localization:
      recall:
        type: lidar.localization_map_recall

        map_version: "multi_exp_point_map"

        visualize: true

      icp:
        type: lidar.localization_icp_v2

        use_pose_prior: false

        # ICP parameters
        num_threads: 4
        first_num_steps: 3
        initial_max_iter: 10
        initial_num_samples: 12000
        initial_max_pairing_dist: 1.5
        initial_max_planar_dist: 1.0
        refined_max_iter: 50
        refined_num_samples: 12000
        refined_max_pairing_dist: 1.0
        refined_max_planar_dist: 0.3

        min_matched_ratio: 0.3

      memory:
        type: "graph_mem_manager"
        vertex_life_span: 5
        window_size: 3
